compiler/fir/resolve/src/org/jetbrains/kotlin/fir/types/ConeTypeCompatibilityChecker.kt

```diff
@@ -0,0 +1,410 @@
+ * parameters bound to these base types: `T1` in `ArrayList<T1>` and `T2` in `List<T2>`. For `T1`, there is only one bound type argument
+ * `Set<String>`, so it's good. For `T2`, there are two bound type arguments: `Set<String>` and `HashSet<Int>`. Now the checker recursively
+ * checks whether these two types are compatible. Again, it first checks the base type `Set` and `HashSet`, and it finds no problem since
+ * `HashSet <: Set`. Finally, checks the type arguments `String` and `Int` that are bound to `T` in `Set<T>`. They are incompatible since
+ * `String` and `Int` are unrelated classes.
+ *
+ * The above example only goes over covariant type arguments. For contravariant types, the checker simply checks whether the range formed by
+ * covariant and contravariant bounds is empty. For example, a range like `[Collection, List]` is empty and hence invalid because `List` is
+ * not a super class/interface of `Collection`
+ */
+object ConeTypeCompatibilityChecker {
+    private val ConeInferenceContext.mimicOldFrontendBehavior: Boolean
+        get() = !session.languageVersionSettings.supportsFeature(LanguageFeature.ExtendedTypeCompatibilityChecker)
+
+    fun Collection<ConeKotlinType>.areCompatible(ctx: ConeInferenceContext): Boolean {
+        // If all types are nullable, then `null` makes the given types compatible.
+        if (all { with(ctx) { it.isNullableType() } }) return true
+        // Next can simply focus on the type hierarchy and don't need to worry about nullability.
+        return ctx.areCompatible(flatMap { it.collectUpperBounds() }.toSet(), emptySet())
+    }
+
+    private fun ConeInferenceContext.areCompatible(upperBounds: Set<ConcreteType>, lowerBounds: Set<ConcreteType>): Boolean {
+        val upperBoundConcreteClassToTypes: Map<ConcreteClass, List<ConcreteType>> =
+            upperBounds.groupBy { it.toConcreteClass(this) }.filterKeys { it != null }.cast()
+
+        // Map all types to the upper bound class in the class hierarchy and check the leaf classes. We can know the types are incompatible
+        // if
+        //   * two or more are not interfaces, or
+        //   * one class is final and there are more than one distinct interfaces
+
+        // Put the upper bound classes into the class hierarchy and count how many subclasses are there for each encountered class
+        val subClassCounts: MutableMap<ConcreteClass, AtomicInteger> = mutableMapOf()
+        val queue = ArrayDeque(upperBoundConcreteClassToTypes.keys)
+        while (queue.isNotEmpty()) {
+            for (superConcreteClass in queue.removeFirst().getSuperConcreteClasses(this)) {
+                subClassCounts.computeIfAbsent(superConcreteClass) {
+                    queue.addLast(it)
+                    AtomicInteger()
+                }.incrementAndGet()
+            }
+        }
+
+        val leafClassesOrInterfaces = upperBoundConcreteClassToTypes.keys.filter { subClassCounts[it]?.get() ?: 0 == 0 }
```

Comment:
I didn't use default value because of the usage of `computeIfAbsent` at line 77. I think using both default value and `computeIfAbsent` can be confusing because line 77 cares about the presence of a key, which is typically not the case when one uses a default value.